### PR TITLE
Internationalize units and currency

### PR DESCRIPTION
**Pull request by JonathanDCohen**
Edit: Currently unit tests fail on Travis but pass on my machine.  For some reason no translation occurs on Travis' servers.
[Demo video](https://youtu.be/3fzmQj2UkY4)

This required updating to Typescript 4 and adding compiler changes for the server manager in order to support newer features of `Intl`.

This change also fixes it so that the server city, creation date, cost, server data cap, data limits, etc are no longer stuck to the old language after changing languages.  Some of these fixes only work the first time you switch languages, but not if you switch again.  I have no idea why that is, but quitting and restarting the app still gets everything back in sync.

This change un-broke the web app tests by correctly building the digitalocean install script to avoid import errors.

### CODE REVIEW

File: src/server_manager/web_app/ui_components/outline-server-view.js [hunk 1]
```diff
@@ -801,43 +803,61 @@ export class ServerView extends DirMixin(PolymerElement) {
     }
   }
 
-  _formatBytesTransferred(numBytes, emptyValue = '') {
+  _makeReadableDataAmount(amount, numDecimals) {
+    const readable = byte_size(amount, {precision: numDecimals});
+    return {
+      amount: readable.value,
+      // Make the unit singular
+      unit: readable.long.slice(0, -1)
```
Comment (fortuna) [inline src/server_manager/web_app/ui_components/outline-server-view.js:811]: I don't think we should rely on this. `long` is supposed to be a huma-readable string. Will it always be an English string? It may change too. We should use the unit only, and map to Intl unit.

Alternatively, let's implement our own bytes formatter. The logic is quite simple.

File: src/server_manager/web_app/ui_components/outline-server-view.js [hunk 2]
```diff
@@ -801,43 +803,61 @@ export class ServerView extends DirMixin(PolymerElement) {
...
-    if (!formattedTransfer) {
+    if (!numBytes) {
       return emptyValue;
     }
-    const value = parseFloat(formattedTransfer);
-    return !!value ? value : emptyValue;
+    const data = this._makeReadableDataAmount(numBytes, this._getFormatPrecision(numBytes));
+    const parts = i18n.makeUnitFormatter(data.unit).formatToParts(data.amount);
+    // Assumes that the last two parts are a space and the unit
+    return parts.slice(0, -2).map(part => part.value).join('');
```
Comment (fortuna) [inline src/server_manager/web_app/ui_components/outline-server-view.js:851]: You can't assume that. Look for `type === 'unit'` instead.

File: src/server_manager/web_app/ui_components/outline-server-view.js [hunk 3]
```diff
@@ -801,43 +803,61 @@ export class ServerView extends DirMixin(PolymerElement) {
...
-    if (!units || units.length < 0) {
+    if (!numBytes) {
       return emptyValue;
     }
-    return units.pop();
+    const data = this._makeReadableDataAmount(numBytes, 0);
+    return i18n.formattedUnit(data.unit, this.language);
   }
 
   _getFormattedTransferredValue(numBytes, emptyValue = '') {
```
Comment (fortuna) [inline src/server_manager/web_app/ui_components/outline-server-view.js:844]: Let's replace this and get rid of bytes_size with our own logic to avoid the hacks to extract the unit and put it back.

File: src/server_manager/web_app/ui_components/outline-server-view.js [hunk 4]
```diff
@@ -801,43 +803,61 @@ export class ServerView extends DirMixin(PolymerElement) {
...
+  _getFormatPrecision(numBytes) {
+    if (numBytes >= 10 ** 9) {
+      return 2;
+    } else if (numBytes >= 10 ** 6) {
+      return 1;
+    }
+    return 0;
+  }
+
+  _formatBytesTransferred(numBytes, language, emptyValue = '') {
```
Comment (fortuna) [inline src/server_manager/web_app/ui_components/outline-server-view.js:824]: Let's get rid of _formatBytesTransferred and _getFormattedTransferredUnit and:

- Create properties for the value and the unit to use in the HTML
- Add a setter for `totalInboundBytes` and remove setServerTransferredData. The setter should set the two new properties.

The two new properties should be set by extracting from the parts returned by the formatter.

This way you format the data only once and we can see all the logic in one place. It's a bit hard to follow now.

File: src/server_manager/web_app/i18n_formatting.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+*/
+
+// Utility functions for internationalizing units and currency
+
+export function makeUnitFormatter(unit: string, language: string) {
```
Comment (fortuna) [inline src/server_manager/web_app/i18n_formatting.ts:19]: Let's replace this with our own version of the of byte_size instead.

Also, code that is not application-specific should live in `infrastructure/`

File: src/server_manager/web_app/app.ts [hunk 6]
```diff
@@ -1254,6 +1255,9 @@ export class App {
     this.appRoot.setLanguage(languageCode, languageDir);
     document.documentElement.setAttribute('dir', languageDir);
     window.localStorage.setItem('overrideLanguage', languageCode);
+    this.appRoot.addEventListener(
```
Comment (fortuna) [inline src/server_manager/web_app/app.ts:1258]: Every time you call setAppLAnguage you will add a new Event Listener.
Add that in the constructor instead.

File: src/server_manager/tsconfig.json [hunk 7]
```diff
@@ -10,5 +10,8 @@
     // FIXME: these tests fail with a runtime error because app.ts depends on
     // polymer, which targets the browser and uses ES6 imports.
     "web_app/app.spec.ts"
-  ]
+  ],
+  "compilerOptions": {
```
Comment (fortuna) [inline src/server_manager/tsconfig.json:14]: Is this really needed? How come this was working before? I thought webpack would take care of those things.
/cc @alalamav

File: src/server_manager/web_app/app.ts [hunk 8]
```diff
@@ -853,6 +853,7 @@ export class App {
     view.isAccessKeyDataLimitEnabled = !!view.accessKeyDataLimit;
     view.showFeatureMetricsDisclaimer = selectedServer.getMetricsEnabled() &&
         !selectedServer.getAccessKeyDataLimit() && !hasSeenFeatureMetricsNotification();
+    view.language = this.appRoot.language;
```
Comment (fortuna) [inline src/server_manager/web_app/app.ts:856]: Specify this binding on the HTML instead

File: package.json [hunk 9]
```diff
@@ -7,7 +7,7 @@
     "husky": "^1.3.1",
     "jasmine": "^3.5.0",
     "tslint": "^5.9.1",
-    "typescript": "^3.8.3"
+    "typescript": "^4"
```
Comment (fortuna) [inline package.json:10]: Could you clarify why we need TS 4?

Comment (fortuna) [reply package.json:10]: Is it required in order to use ES2020?

File: src/server_manager/tsconfig.json (hunk 7 above)
Comment (fortuna) [reply src/server_manager/tsconfig.json:14]: What do you need from ES2020?

You shouldn't have to specify multiple ES versions.

**Review by fortuna — CHANGES_REQUESTED** [review]
Thanks for cleaning up the units.

File: package.json (hunk 9 above)
Comment (JonathanDCohen) [reply package.json:10]: [This `tsc` bug](https://github.com/microsoft/TypeScript/pull/38013) was merged to master on June 18th.  Typescript 3.9 was [released May 12th](https://devblogs.microsoft.com/typescript/announcing-typescript-3-9/) and Typescript 4/0 was released [August 12th](https://devblogs.microsoft.com/typescript/announcing-typescript-4-0/).  So we need Typescript 4.0 in order to get that bugfix.  Alternatively we could cast all of our uses of `Intl.NumberFormatter` but that seems like a worse idea than just upgrading our Typescript version.

File: src/server_manager/tsconfig.json (hunk 7 above)
Comment (JonathanDCohen) [reply src/server_manager/tsconfig.json:14]: It pulls in [This tsc bugfix](https://github.com/microsoft/TypeScript/pull/38013#issuecomment-646314376) which lives, as per this comment (and also I confirmed by checking) in `es2020.intl.d.ts`.  This is how we pull in that version of `Intl`.

File: src/server_manager/web_app/app.ts (hunk 6 above)
Comment (JonathanDCohen) [reply src/server_manager/web_app/app.ts]: Done in 79fb1ea

File: src/server_manager/web_app/i18n_formatting.ts (hunk 5 above)
Comment (JonathanDCohen) [reply src/server_manager/web_app/i18n_formatting.ts]: Done in 79fb1ea

File: src/server_manager/web_app/ui_components/outline-server-view.js (hunk 4 above)
Comment (JonathanDCohen) [reply src/server_manager/web_app/ui_components/outline-server-view.js:805]: I didn't completely get rid of _formatBytesTransferred, as it also provides default values in a number of places.  I did reduce use of it and have it call into the i18n library.  79fb1ea

File: src/server_manager/web_app/ui_components/outline-server-view.js (hunk 3 above)
Comment (JonathanDCohen) [reply src/server_manager/web_app/ui_components/outline-server-view.js]: Done in 79fb1ea

File: src/server_manager/web_app/ui_components/outline-server-view.js (hunk 2 above)
Comment (JonathanDCohen) [reply src/server_manager/web_app/ui_components/outline-server-view.js]: Done in 79fb1ea

File: src/server_manager/web_app/ui_components/outline-server-view.js (hunk 1 above)
Comment (JonathanDCohen) [reply src/server_manager/web_app/ui_components/outline-server-view.js]: Done.  I got rid of our use of byte-size completely.

File: src/server_manager/tsconfig.json (hunk 7 above)
Comment (fortuna) [reply src/server_manager/tsconfig.json:14]: Let's install the Format.JS polyfill instead: https://formatjs.io/docs/polyfills/intl-numberformat

This way we don't need to make any changes to the existing build. You will get the right types, without having to change tsconfig or the TS version.

**Review by fortuna — CHANGES_REQUESTED** [review]

File: src/server_manager/infrastructure/data_formatting.ts [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+}
+
+export function getFormattedDataAmountParts(amount: number, language: string) {
+  const params = getDataFormattingParams(amount);
+  const parts = makeUnitFormatter(language, params).formatToParts(params.value);
+  const isUnit = (part: Intl.NumberFormatPart) => (part as {type: string}).type === 'unit';
+  const unit = parts.find(isUnit).value;
+  return {
+    value: parts.filter((part) => !isUnit(part)).map(part => part.value).join('').trim(),
+    unit: unit === 'byte' ? 'B' : unit
```
Comment (fortuna) [inline src/server_manager/infrastructure/data_formatting.ts:60]: Rename `unit:` to `unit_text:` to not confuse with the unit id.

File: src/server_manager/infrastructure/data_formatting.spec.ts [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+  limitations under the License.
+*/
+
+import * as i18n from './data_formatting';
+
+// Jasmine with Node only supports 'en'
+if (Intl.NumberFormat.supportedLocalesOf(['en', 'fr', 'fa', 'es', 'ru']).length === 5) {
+  describe('getFormattedDataAmountParts', () => {
+    it('extracts the unit string and value separately', () => {
+      const english = i18n.getFormattedDataAmountParts(0, 'en');
```
Comment (fortuna) [inline src/server_manager/infrastructure/data_formatting.spec.ts:23]: Could you add an example where the value is 2 bytes?

File: src/server_manager/web_app/karma.conf.js [hunk 12]
```diff
@@ -18,15 +18,22 @@ const baseConfig = makeConfig({
   defaultMode: 'development'
 });
 
+const test_patterns = [
+  '**/*.spec.ts',
+  // We need to test data_formatting in a browser context
+  '../infrastructure/data_formatting.spec.ts'
```
Comment (fortuna) [inline src/server_manager/web_app/karma.conf.js:24]: Why do  we need to test the formatting in a browser context?  It doesn't use the DOM.

File: src/server_manager/infrastructure/data_formatting.ts [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+    return {value: numBytes / KILOBYTE, unit: 'kilobyte', decimalPlaces: 0};
+  }
+  return {value: numBytes, unit: 'byte', decimalPlaces: 0};
+}
+
+function makeUnitFormatter(language: string, params: FormatParams) {
+  const options: Intl.NumberFormatOptions = {
+    style: 'unit',
+    unit: params.unit,
+    maximumFractionDigits: params.decimalPlaces
```
Comment (fortuna) [inline src/server_manager/infrastructure/data_formatting.ts:47]: Maybe force `unitDisplay == 'short'`, for clarity?

File: src/server_manager/web_app/ui_components/outline-server-view.js [hunk 14]
```diff
@@ -801,43 +802,22 @@ export class ServerView extends DirMixin(PolymerElement) {
...
-    if (!formattedTransfer) {
-      return emptyValue;
+  _formatMonthlyCost(monthlyCost) {
+    if (!monthlyCost) {
+      return '';
     }
-    const value = parseFloat(formattedTransfer);
-    return !!value ? value : emptyValue;
+    return new Intl
+        .NumberFormat(this.language, {style: 'currency', currency: 'USD', currencyDisplay: 'code'})
```
Comment (fortuna) [inline src/server_manager/web_app/ui_components/outline-server-view.js:819]: Is this a cheap constructor? Doesn't it load some resources? I wonder if  we should cache the formatter.

File: src/server_manager/web_app/ui_components/outline-server-view.js [hunk 15]
```diff
@@ -665,6 +660,8 @@ export class ServerView extends DirMixin(PolymerElement) {
        */
       this.myConnection = null;
       this.totalInboundBytes = 0;
+      this.inboundByesUnit = 'B';
```
Comment (fortuna) [inline src/server_manager/web_app/ui_components/outline-server-view.js:663]: We should either leave this empty, or set it according to the language.
Maybe you set those fields on the connected callback, or when the language is set.

File: src/server_manager/web_app/app.ts [hunk 16]
```diff
@@ -293,6 +294,9 @@ export class App {
       this.handleShowServerRequested(event.detail.displayServerId);
     });
 
+    this.appRoot.addEventListener(
```
[the remark this answers is not in the document]
Comment (fortuna) [reply src/server_manager/web_app/app.ts:297]: But see my other comment on moving this event handling to setLanguage.

**Review by fortuna — CHANGES_REQUESTED** [review]

File: src/server_manager/tsconfig.json (hunk 7 above)
Comment (JonathanDCohen) [reply src/server_manager/tsconfig.json:14]: Using the polyfill requires dynamically importing a new module for each language change.  This runs into an error: "Dynamic imports are only supported when the '--module' flag is set to 'es2020', 'esnext', 'commonjs', 'amd', 'system', or 'umd'". So we will have to set these flags anyways to use the polyfill.  Because of this I chose to use the system supported version since this is only used in browser code, not in Node.

File: src/server_manager/infrastructure/data_formatting.ts (hunk 10 above)
Comment (JonathanDCohen) [reply src/server_manager/infrastructure/data_formatting.ts]: Done in 8c1a131cb12e6b52de6008df1eaf5526011fa487

File: src/server_manager/infrastructure/data_formatting.ts [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+  };
+  const out = new Intl.NumberFormat(language, options);
+  return out;
+}
+
+export function getFormattedDataAmountParts(amount: number, language: string) {
+  const params = getDataFormattingParams(amount);
+  const parts = makeUnitFormatter(language, params).formatToParts(params.value);
+  const isUnit = (part: Intl.NumberFormatPart) => (part as {type: string}).type === 'unit';
+  const unit = parts.find(isUnit).value;
```
[the remark this answers is not in the document]
Comment (JonathanDCohen) [reply src/server_manager/infrastructure/data_formatting.ts]: Done in 8c1a131cb12e6b52de6008df1eaf5526011fa487

File: src/server_manager/infrastructure/data_formatting.ts (hunk 13 above)
Comment (JonathanDCohen) [reply src/server_manager/infrastructure/data_formatting.ts]: Done in 8c1a131cb12e6b52de6008df1eaf5526011fa487

File: src/server_manager/infrastructure/data_formatting.spec.ts (hunk 11 above)
Comment (JonathanDCohen) [reply src/server_manager/infrastructure/data_formatting.spec.ts]: Done in 8c1a131cb12e6b52de6008df1eaf5526011fa487

File: src/server_manager/web_app/ui_components/outline-server-settings.js [hunk 18]
```diff
@@ -356,5 +359,20 @@ Polymer({
     const port = Number(value);
     const valid = !Number.isNaN(port) && port >= 1 && port <= 65535 && Number.isInteger(port);
     return valid ? '' : this.localize('error-keys-port-bad-input');
+  },
+
+  _getInternationalizedUnit(bytesAmount, language) {
+    // This happens during app startup before we set the language
```
Comment (mpmcroy) [inline src/server_manager/web_app/ui_components/outline-server-settings.js:365]: I've run into these kind of issues a few times too.

My approach has been to try to set a sensible default and always maintain an invariant that the value is not null. That allowed me to remove the `null|undefined` checks.

What are your thoughts?

File: src/server_manager/web_app/ui_components/outline-server-view.js [hunk 19]
```diff
@@ -596,7 +588,7 @@ export class ServerView extends DirMixin(PolymerElement) {
         serverManagementApiUrl: String,
         serverPortForNewAccessKeys: Number,
         isAccessKeyPortEditable: {type: Boolean},
-        serverCreationDate: String,
+        serverCreationDate: {type: Object, value: null},  // type: Date
```
Comment (mpmcroy) [inline src/server_manager/web_app/ui_components/outline-server-view.js:591]: Can you add a JS Doc type comment to the constructor instead?

File: src/server_manager/web_app/data_formatting.spec.ts [hunk 20]
```diff
@@ -0,0 +1,52 @@
...
+
+      const farsi = i18n.getFormattedDataAmountParts(133.5 * 10 ** 6, 'fa');
+      expect(farsi.unit).toEqual('مگابایت');
+      expect(farsi.value).toEqual('۱۳۳٫۵');
+    });
+  });
+
+  describe('formatBytes', () => {
+    it('Formats data amounts', () => {
+      expect(i18n.formatBytes(10 * 10 ** 9, 'en')).toEqual('10 GB');
```
Comment (mpmcroy) [inline src/server_manager/web_app/data_formatting.spec.ts:43]: Can you add a KB test so that we cover that code path?

File: src/server_manager/web_app/ui_components/app-root.js [hunk 21]
```diff
@@ -561,15 +561,29 @@ export class AppRoot extends mixinBehaviors
    * @param {string} language
    * @param {string} direction
    */
-  setLanguage(language, direction) {
+  async setLanguage(language, direction) {
+    const out = new Promise((resolve, reject) => {
+      let successHandler, failureHandler;
+      successHandler = () => {
+        this.removeEventListener('app-localize-resources-error', failureHandler);
```
Comment (mpmcroy) [inline src/server_manager/web_app/ui_components/app-root.js:568]: Can you explain the event listeners here? It seems like just wrapping the setLanguage logic in a promise should be enough.

File: src/server_manager/web_app/app.ts [hunk 22]
```diff
@@ -1100,10 +1101,15 @@ export class App {
...
-  private setAppLanguage(languageCode: string, languageDir: string) {
-    this.appRoot.setLanguage(languageCode, languageDir);
-    document.documentElement.setAttribute('dir', languageDir);
-    window.localStorage.setItem('overrideLanguage', languageCode);
+  private async setAppLanguage(languageCode: string, languageDir: string) {
+    try {
+      await this.appRoot.setLanguage(languageCode, languageDir);
+      document.documentElement.setAttribute('dir', languageDir);
+      window.localStorage.setItem('overrideLanguage', languageCode);
+      this.appRoot.showServerView();
```
Comment (mpmcroy) [inline src/server_manager/web_app/app.ts:1109]: I don't think we should call `showServerView` here. It might be confusing to the user if the app changes screens.

**Review by mpmcroy — COMMENTED** [review]
The internationalization looks good! 

And thanks for fixing the localization and karma tests 🎉

File: src/server_manager/web_app/ui_components/app-root.js (hunk 21 above)
Comment (mpmcroy) [reply src/server_manager/web_app/ui_components/app-root.js:573]: Oh thanks. Yes, that clears it up!

File: src/server_manager/web_app/ui_components/outline-server-settings.js [hunk 23]
```diff
@@ -356,5 +359,20 @@ Polymer({
...
     return valid ? '' : this.localize('error-keys-port-bad-input');
+  },
+
+  _getInternationalizedUnit(bytesAmount, language) {
+    return formatBytesParts(bytesAmount, language).unit;
+  },
+
+  _getTranslatedDate(language, date) {
+    // We can't use a default Date object -- it still shows up as null, I'm not sure the lifetime
+    // properties when you pass an object into Polymer.
```
Comment (mpmcroy) [inline src/server_manager/web_app/ui_components/outline-server-settings.js:370]: I wonder if this happens because we set the default value in the Polymer property configuration. I think this is one of the only files that sets defaults that way.

Does it work if you set `serverCreationDate` in the constructor?

File: src/server_manager/web_app/ui_components/outline-server-settings.js (hunk 18 above)
Comment (mpmcroy) [reply src/server_manager/web_app/ui_components/outline-server-settings.js]: That seems like a good default to me. I suspect the user will not even see the default value since it's not the default tab.

**Review by mpmcroy — APPROVED** [review]
I didn't take a look at the Typescript 4/ES 2020 part since I think @fortuna is better suited to review that, but the rest looks 🔥 🔥 🔥 

LGTM!

File: src/server_manager/web_app/ui_components/outline-server-settings.js (hunk 23 above)
Comment (JonathanDCohen) [reply src/server_manager/web_app/ui_components/outline-server-settings.js:370]: I would guess that it does, but I don't want to change this file to use a constructor in this PR, it's too much going on already and it's not important enough to keep delaying the per key data limits ui for

File: src/server_manager/web_app/karma.conf.js [hunk 24]
```diff
@@ -18,15 +18,22 @@ const baseConfig = makeConfig({
   defaultMode: 'development'
 });
 
+const test_patterns = [
+  '**/*.spec.ts',
+  // We need to test data_formatting in a browser context
+  '../web_app/data_formatting.spec.ts'
```
Comment (fortuna) [inline src/server_manager/web_app/karma.conf.js:24]: What's the working directory here? isn't it we_app already?

File: src/server_manager/web_app/ui_components/outline-server-settings.js [hunk 25]
```diff
@@ -356,5 +359,20 @@ Polymer({
     const port = Number(value);
     const valid = !Number.isNaN(port) && port >= 1 && port <= 65535 && Number.isInteger(port);
     return valid ? '' : this.localize('error-keys-port-bad-input');
+  },
+
+  _getInternationalizedUnit(bytesAmount, language) {
+    return formatBytesParts(bytesAmount, language).unit;
+  },
+
+  _getTranslatedDate(language, date) {
```
Comment (fortuna) [inline src/server_manager/web_app/ui_components/outline-server-settings.js:368]: A name like `formatDate` would be more appropriate. There's no translation (from one language to another) going on.

File: src/server_manager/web_app/ui_components/outline-server-settings.js [hunk 26]
```diff
@@ -274,10 +276,11 @@ Polymer({
         {type: Boolean, value: false},  // Whether the server supports data limits.
     showFeatureMetricsDisclaimer: {type: Boolean, value: false},
     isHostnameEditable: {type: Boolean, value: true},
-    serverCreationDate: {type: String, value: null},
+    serverCreationDate: {type: Object, value: null},
```
Comment (fortuna) [inline src/server_manager/web_app/ui_components/outline-server-settings.js:279]: I think you can use Date here

File: src/server_manager/web_app/ui_components/outline-server-view.js [hunk 27]
```diff
@@ -782,8 +777,20 @@ export class ServerView extends DirMixin(PolymerElement) {
     this.dispatchEvent(makePublicEvent('RemoveAccessKeyRequested', {accessKeyId: accessKey.id}));
   }
 
-  setServerTransferredData(totalBytes) {
-    this.totalInboundBytes = totalBytes;
+  _getInboundBytesUnit(totalBytes, language) {
```
Comment (fortuna) [inline src/server_manager/web_app/ui_components/outline-server-view.js:780]: Perhaps `_formatInboundBytesUnit` and `_formatInboundBytesValue` is clearer.

The unit is bytes, so "get unit" isn't quite accurate.

File: src/server_manager/web_app/ui_components/outline-server-view.js [hunk 28]
```diff
@@ -801,43 +808,22 @@ export class ServerView extends DirMixin(PolymerElement) {
...
-      return emptyValue;
-    }
-    return units.pop();
-  }
-
-  _getFormattedTransferredValue(numBytes, emptyValue = '') {
-    const formattedTransfer = this._formatBytesTransferred(numBytes);
-    if (!formattedTransfer) {
-      return emptyValue;
+  _formatMonthlyCost(monthlyCost) {
```
Comment (fortuna) [inline src/server_manager/web_app/ui_components/outline-server-view.js:820]: This should take the language as input, so it's recalculated if the language changes.

File: src/server_manager/web_app/ui_components/app-root.js [hunk 29]
```diff
@@ -561,15 +561,34 @@ export class AppRoot extends mixinBehaviors
...
+        const alignDir = direction === 'ltr' ? 'left' : 'right';
+        this.$.appDrawer.align = alignDir;
+        this.$.sideBar.align = alignDir;
+        this.language = language;
+        resolve();
+      };
+      failureHandler = (err) => {
```
Comment (fortuna) [inline src/server_manager/web_app/ui_components/app-root.js:580]: The input here is an event, not an error, but I think it will be empty: https://github.com/PolymerElements/app-localize-behavior/blob/e796bfa6876498608ab2d0d63c89bffece80c0da/app-localize-behavior.js#L327

File: src/server_manager/web_app/ui_components/app-root.js [hunk 30]
```diff
@@ -561,15 +561,34 @@ export class AppRoot extends mixinBehaviors
...
+        this.removeEventListener('app-localize-resources-loaded', successHandler);
+        reject(err);
```
Comment (fortuna) [inline src/server_manager/web_app/ui_components/app-root.js:582]: We need to create an Error, since err is not one.

File: src/server_manager/web_app/ui_components/app-root.js (hunk 21 above)
Comment (fortuna) [reply src/server_manager/web_app/ui_components/app-root.js:573]: This is a bit hard to read. Can you provide a method that all it does is load the resources and return the completion promise?

Then the setLanguage can read almost like it was before, except that it will be async and have an `await loadResources()`.

I'm not sure about the name. Perhaps loadResourcesAsync? loadResourcesPromise?

File: src/server_manager/tsconfig.json (hunk 7 above)
Comment (fortuna) [reply src/server_manager/tsconfig.json:14]: If we are assuming the code is running on an ES2020 environment, then set the target to ES2020 and revert the lib option. Otherwise it's very confusing.

**Review by fortuna — CHANGES_REQUESTED** [review]

File: src/server_manager/web_app/data_formatting.spec.ts (hunk 20 above)
Comment (JonathanDCohen) [reply src/server_manager/web_app/data_formatting.spec.ts:47]: It's in https://github.com/Jigsaw-Code/outline-server/pull/780/commits/49ce60d8d96a8d1d434992c60e01f04c2a37af65#diff-858c37f4a836374e98fb54a760e5a4099b33a6ae04b41c2079b60c588eeb492cR31, I'm not sure why the view isn't showing it

File: src/server_manager/web_app/ui_components/app-root.js (hunk 21 above)
Comment (JonathanDCohen) [reply src/server_manager/web_app/ui_components/app-root.js:574]: Fixed after discussion online in 9d11943e51379020fda0b9357457e185e844dfc3

File: src/server_manager/web_app/app.ts [hunk 31]
```diff
@@ -1100,10 +1101,14 @@ export class App {
     });
   }
 
-  private setAppLanguage(languageCode: string, languageDir: string) {
-    this.appRoot.setLanguage(languageCode, languageDir);
-    document.documentElement.setAttribute('dir', languageDir);
-    window.localStorage.setItem('overrideLanguage', languageCode);
+  private async setAppLanguage(languageCode: string, languageDir: string) {
+    try {
```
[the remark this answers is not in the document]
Comment (JonathanDCohen) [reply src/server_manager/web_app/app.ts:1105]: The only caller is earlier in the file where we set this as the event listener for 'SetLanguageRequested'.  I think the error handling is more appropriate here than in the wiring code.

File: src/server_manager/tsconfig.json (hunk 7 above)
Comment (JonathanDCohen) [reply src/server_manager/tsconfig.json]: After extensive offline work, we decided to get rid of the compiler options and cast away the type issues.  I added a big warning at the top of the file that we're using an implicit ES2020 dependency for browser code here.

File: src/server_manager/web_app/data_formatting.ts [hunk 32]
```diff
@@ -0,0 +1,75 @@
...
+    return {value: numBytes / GIGABYTE, unit: 'gigabyte', decimalPlaces: 2};
+  } else if (numBytes >= MEGABYTE) {
+    return {value: numBytes / MEGABYTE, unit: 'megabyte', decimalPlaces: 1};
+  } else if (numBytes >= KILOBYTE) {
+    return {value: numBytes / KILOBYTE, unit: 'kilobyte', decimalPlaces: 0};
+  }
+  return {value: numBytes, unit: 'byte', decimalPlaces: 0};
+}
+
+function makeUnitFormatter(language: string, params: FormatParams) {
```
[the remark this answers is not in the document]
Comment (JonathanDCohen) [reply src/server_manager/web_app/data_formatting.ts]: I used `makeDataAmountFormatter`

File: src/server_manager/web_app/ui_components/app-root.js (hunk 29 above)
Comment (JonathanDCohen) [reply src/server_manager/web_app/ui_components/app-root.js]: Done in 9d11943e51379020fda0b9357457e185e844dfc3

File: src/server_manager/web_app/ui_components/app-root.js (hunk 30 above)
Comment (JonathanDCohen) [reply src/server_manager/web_app/ui_components/app-root.js]: Done in 9d11943e51379020fda0b9357457e185e844dfc3

File: src/server_manager/web_app/ui_components/outline-server-settings.js (hunk 25 above)
Comment (JonathanDCohen) [reply src/server_manager/web_app/ui_components/outline-server-settings.js]: Done in 435c5a5a56dab1230b3f11762438144ba3f5d7cf

File: src/server_manager/web_app/ui_components/outline-server-view.js (hunk 27 above)
Comment (JonathanDCohen) [reply src/server_manager/web_app/ui_components/outline-server-view.js]: Done in 435c5a5a56dab1230b3f11762438144ba3f5d7cf

File: src/server_manager/web_app/ui_components/outline-server-view.js (hunk 28 above)
Comment (JonathanDCohen) [reply src/server_manager/web_app/ui_components/outline-server-view.js]: Done in 9d11943e51379020fda0b9357457e185e844dfc3.  I also added the language dependency to the server location name so that will update as well

File: src/server_manager/web_app/data_formatting.spec.ts [hunk 33]
```diff
@@ -0,0 +1,65 @@
...
+      const french = i18n.formatBytesParts(1.5 * 10 ** 9, 'fr');
+      expect(french.unit).toEqual('Go');
+      expect(french.value).toEqual('1,5');
+
+      const farsi = i18n.formatBytesParts(133.5 * 10 ** 6, 'fa');
+      expect(farsi.unit).toEqual('مگابایت');
+      expect(farsi.value).toEqual('۱۳۳٫۵');
+    });
+  }
+});
```
Comment (fortuna) [inline src/server_manager/web_app/data_formatting.spec.ts:47]: This looks good. An alternative is to mark the test explicitly as skipped on Node, so it shows on the output:

```suggestion
  let itBrowser = it;
  if (process?.versions?.node) {
    itBrowser = xit;
  }
  itBrowser('extracts the unit string and value separately', () => {
    const english = i18n.formatBytesParts(0, 'en');
    expect(english.unit).toEqual('B');
    expect(english.value).toEqual('0');
    
    const spanish = i18n.formatBytesParts(2, 'es');
    expect(spanish.unit).toEqual('B');
    expect(spanish.value).toEqual('2');

    const russian = i18n.formatBytesParts(3000, 'ru');
    expect(russian.unit).toEqual('кБ');
    expect(russian.value).toEqual('3');

    const french = i18n.formatBytesParts(1.5 * 10 ** 9, 'fr');
    expect(french.unit).toEqual('Go');
    expect(french.value).toEqual('1,5');

    const farsi = i18n.formatBytesParts(133.5 * 10 ** 6, 'fa');
    expect(farsi.unit).toEqual('مگابایت');
    expect(farsi.value).toEqual('۱۳۳٫۵');
  });
});
```

Up to you

File: src/server_manager/web_app/data_formatting.spec.ts (hunk 20 above)
Comment (fortuna) [reply src/server_manager/web_app/data_formatting.spec.ts:56]: It seems the KB test was added to the `formatBytesParts` above.

**Review by fortuna — APPROVED** [review]
Thanks for making this change. Our unit formatting is a lot more robust now, thanks to you.

Please make sure the test is passing before submitting. It seems to be broken.

File: src/server_manager/web_app/data_formatting.spec.ts (hunk 33 above)
Comment (JonathanDCohen) [reply src/server_manager/web_app/data_formatting.spec.ts:47]: That's a cool idea, but I'm going to keep it the way I wrote it because it tests the runtime check which keeps us from putting the code in Node

**Review by mpmcroy — APPROVED** [review]
🔥
